src/hash.h

```diff
@@ -160,6 +160,42 @@ class CHashWriter
     }
 };
 
+/** Reads data from an underlying stream, while hashing the read data. */
+template<typename Source>
+class CHashVerifier : public CHashWriter
+{
+private:
+    Source* source;
+
+public:
+    CHashVerifier(Source* source_) : CHashWriter(source_->GetType(), source_->GetVersion()), source(source_) {}
```

Comment:
I don't like using references for objects that persist. This risks calling CHashVerifier(function_that_returns_a_Source()).